neuro_extras/data/operations.py

```diff
@@ -0,0 +1,179 @@
+        self.source_type = UrlType.get_type(source)
+        self.destination_type = UrlType.get_type(destination)
+        self.source_filename = get_filename_from_url(source)
+        self.destination_filename = get_filename_from_url(destination)
+        self._ensure_can_execute()
+
+    def _ensure_can_execute(self) -> None:
+        """Raise exception if operation is unsupported
+
+        Possible reasons:
+        - at least one of the source or destination is of unsupported type
+        - pair (source, destination) is explicitly forbidden
+        """
+        if self.source_type == UrlType.UNSUPPORTED:
+            raise ValueError(f"Unsupported source: {self.source}")
+        if self.destination_type == UrlType.UNSUPPORTED:
+            raise ValueError(f"Unsupported destination: {self.destination}")
+        is_forbidden_combination = any(
+            self.source_type == source and self.destination_type == destination
+            for (source, destination) in CopyOperation.get_forbidden_combinations()
+        )
+        if is_forbidden_combination:
+            raise ValueError(
+                f"Copy from {self.source_type.name} to "
+                f"{self.destination_type.name} is unsupported"
+            )
+        else:
+            logger.info(
+                f"Copy from {self.source_type.name} to "
+                f"{self.destination_type.name} is supported"
+            )
+
+    async def run(self) -> None:
+        """Run copy operation.
+
+        Uses appropriate copier instance, that supports source and destionation
+        """
+        async with get_neuro_client() as neuro_client:
+            with provide_temp_dir() as temp_dir:
+                logger.info("Resolving copier...")
+                copier = _get_copier(
+                    source=self.source,
+                    destination=self.destination,
+                    compress=self.compress,
+                    extract=self.extract,
+                    temp_dir=Path(temp_dir),
+                    neuro_client=neuro_client,
+                    volumes=self.volumes,
+                    env=self.env,
+                    life_span=self.life_span,
+                    preset=self.preset,
+                )
+                logger.info(f"Using {copier.__class__.__name__}")
+                await copier.perform_copy()
+
+    @staticmethod
+    def get_forbidden_combinations() -> List[Tuple[UrlType, UrlType]]:
+        """Get forbidden combinations of source and destination types"""
+        return [
+            (UrlType.CLOUD, UrlType.CLOUD),
+            # TODO: (A.K.) implement platform-to-local and vice-versa
+            # through neuro storage cp
+            (UrlType.PLATFORM, UrlType.LOCAL_FS),
+            (UrlType.LOCAL_FS, UrlType.PLATFORM),
+            (UrlType.STORAGE, UrlType.STORAGE),
+            (UrlType.DISK, UrlType.DISK),
+            (UrlType.SUPPORTED, UrlType.WEB),
+        ]
+
+
+def _get_copier(
```

Comment:
Since there are plans for `TransferOperation`, it makes sense to leave the method inside the global scope. Might move it inside the base `DataOperation` class when working on transfer though

Reply:
got you, thanks for the clarification!